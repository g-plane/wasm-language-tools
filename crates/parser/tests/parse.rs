use codespan_reporting::{
    diagnostic::{Diagnostic, Label},
    files::SimpleFile,
    term::{self, termcolor::Buffer},
};
use insta::{assert_snapshot, glob, Settings};
use std::{fs, path::Path};

#[test]
fn parser_snapshot() {
    glob!("parse/**/*.wat", |path| {
        let input = fs::read_to_string(path).unwrap();
        let (tree, errors) = wat_parser::parse2(&input);
        similar_asserts::assert_eq!(
            tree.to_string(),
            input,
            "syntax tree of '{}' does not match source",
            path.display()
        );

        let file = SimpleFile::new(path.file_name().unwrap().to_str().unwrap(), &input);
        let config = term::Config::default();
        let mut buffer = Buffer::no_color();
        errors
            .into_iter()
            .map(|error| {
                Diagnostic::error()
                    .with_message(error.message.to_string())
                    .with_labels(vec![Label::primary((), error.start..error.end)])
            })
            .for_each(|diagnostic| {
                term::emit(&mut buffer, &config, &file, &diagnostic).unwrap();
            });

        build_settings(path).bind(|| {
            let name = path.file_stem().unwrap().to_str().unwrap();
            assert_snapshot!(
                name,
                format!(
                    "{tree:#?}\n{}",
                    String::from_utf8(buffer.into_inner()).unwrap()
                )
            );
        });
    });
}

fn build_settings(path: &Path) -> Settings {
    let mut settings = Settings::clone_current();
    settings.set_snapshot_path(path.parent().unwrap());
    settings.remove_snapshot_suffix();
    settings.set_prepend_module_to_snapshot(false);
    settings.remove_input_file();
    settings.set_omit_expression(true);
    settings.remove_input_file();
    settings.remove_info();
    settings
}
