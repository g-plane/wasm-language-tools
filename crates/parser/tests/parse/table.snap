---
source: crates/parser/tests/parse.rs
---
ROOT@0..215
  MODULE@0..26
    L_PAREN@0..1 "("
    KEYWORD@1..7 "module"
    WHITESPACE@7..8 " "
    MODULE_FIELD_TABLE@8..25
      L_PAREN@8..9 "("
      KEYWORD@9..14 "table"
      WHITESPACE@14..15 " "
      TABLE_TYPE@15..24
        LIMITS@15..16
          UNSIGNED_INT@15..16 "0"
        WHITESPACE@16..17 " "
        REF_TYPE@17..24
          TYPE_KEYWORD@17..24 "funcref"
      R_PAREN@24..25 ")"
    R_PAREN@25..26 ")"
  WHITESPACE@26..27 "\n"
  MODULE@27..53
    L_PAREN@27..28 "("
    KEYWORD@28..34 "module"
    WHITESPACE@34..35 " "
    MODULE_FIELD_TABLE@35..52
      L_PAREN@35..36 "("
      KEYWORD@36..41 "table"
      WHITESPACE@41..42 " "
      TABLE_TYPE@42..51
        LIMITS@42..43
          UNSIGNED_INT@42..43 "1"
        WHITESPACE@43..44 " "
        REF_TYPE@44..51
          TYPE_KEYWORD@44..51 "funcref"
      R_PAREN@51..52 ")"
    R_PAREN@52..53 ")"
  WHITESPACE@53..54 "\n"
  MODULE@54..82
    L_PAREN@54..55 "("
    KEYWORD@55..61 "module"
    WHITESPACE@61..62 " "
    MODULE_FIELD_TABLE@62..81
      L_PAREN@62..63 "("
      KEYWORD@63..68 "table"
      WHITESPACE@68..69 " "
      TABLE_TYPE@69..80
        LIMITS@69..72
          UNSIGNED_INT@69..70 "0"
          WHITESPACE@70..71 " "
          UNSIGNED_INT@71..72 "0"
        WHITESPACE@72..73 " "
        REF_TYPE@73..80
          TYPE_KEYWORD@73..80 "funcref"
      R_PAREN@80..81 ")"
    R_PAREN@81..82 ")"
  WHITESPACE@82..83 "\n"
  MODULE@83..111
    L_PAREN@83..84 "("
    KEYWORD@84..90 "module"
    WHITESPACE@90..91 " "
    MODULE_FIELD_TABLE@91..110
      L_PAREN@91..92 "("
      KEYWORD@92..97 "table"
      WHITESPACE@97..98 " "
      TABLE_TYPE@98..109
        LIMITS@98..101
          UNSIGNED_INT@98..99 "0"
          WHITESPACE@99..100 " "
          UNSIGNED_INT@100..101 "1"
        WHITESPACE@101..102 " "
        REF_TYPE@102..109
          TYPE_KEYWORD@102..109 "funcref"
      R_PAREN@109..110 ")"
    R_PAREN@110..111 ")"
  WHITESPACE@111..112 "\n"
  MODULE@112..142
    L_PAREN@112..113 "("
    KEYWORD@113..119 "module"
    WHITESPACE@119..120 " "
    MODULE_FIELD_TABLE@120..141
      L_PAREN@120..121 "("
      KEYWORD@121..126 "table"
      WHITESPACE@126..127 " "
      TABLE_TYPE@127..140
        LIMITS@127..132
          UNSIGNED_INT@127..128 "1"
          WHITESPACE@128..129 " "
          UNSIGNED_INT@129..132 "256"
        WHITESPACE@132..133 " "
        REF_TYPE@133..140
          TYPE_KEYWORD@133..140 "funcref"
      R_PAREN@140..141 ")"
    R_PAREN@141..142 ")"
  WHITESPACE@142..143 "\n"
  MODULE@143..175
    L_PAREN@143..144 "("
    KEYWORD@144..150 "module"
    WHITESPACE@150..151 " "
    MODULE_FIELD_TABLE@151..174
      L_PAREN@151..152 "("
      KEYWORD@152..157 "table"
      WHITESPACE@157..158 " "
      TABLE_TYPE@158..173
        LIMITS@158..165
          UNSIGNED_INT@158..159 "0"
          WHITESPACE@159..160 " "
          UNSIGNED_INT@160..165 "65536"
        WHITESPACE@165..166 " "
        REF_TYPE@166..173
          TYPE_KEYWORD@166..173 "funcref"
      R_PAREN@173..174 ")"
    R_PAREN@174..175 ")"
  WHITESPACE@175..176 "\n"
  MODULE@176..214
    L_PAREN@176..177 "("
    KEYWORD@177..183 "module"
    WHITESPACE@183..184 " "
    MODULE_FIELD_TABLE@184..213
      L_PAREN@184..185 "("
      KEYWORD@185..190 "table"
      WHITESPACE@190..191 " "
      TABLE_TYPE@191..212
        LIMITS@191..204
          UNSIGNED_INT@191..192 "0"
          WHITESPACE@192..193 " "
          UNSIGNED_INT@193..204 "0xffff_ffff"
        WHITESPACE@204..205 " "
        REF_TYPE@205..212
          TYPE_KEYWORD@205..212 "funcref"
      R_PAREN@212..213 ")"
    R_PAREN@213..214 ")"
  WHITESPACE@214..215 "\n"
